brainiak/reprsimil/brsa.py

```diff
@@ -30,24 +32,158 @@
+        large values of tau, while half-Cauchy prior only penalizes
+        for very large values of tau.
+        For more information on usage, see description in BRSA class:
+        `.BRSA`
+    """
+    tau2 = (y_invK_y - n_y * tau_range**2
+            + np.sqrt(n_y**2 * tau_range**4 + (2 * n_y + 8)
+                      * tau_range**2 * y_invK_y + y_invK_y**2))\
+        / 2 / (n_y + 2)
+    log_ptau = scipy.stats.halfcauchy.logpdf(
+        tau2**0.5, scale=tau_range)
+    return tau2, log_ptau
+
+
+def Ncomp_BIC_Minka(X):
+    """ This implements the BIC metric written in Minka 2000:
+        Automatic choice of dimensionality for PCA
+        The paper mainly proposed a Laplace approximation of the posterior
+        of the number of components but also provided this BIC approximation.
+        For cases in which feature number is smaller than sample size,
+        there is ambiguity for the author of this code to decide how to
+        set some variables. Empirically the number of components estimated
+        turn to be very large by the Laplace approximation method. The BIC
+        method appear to perform well in a limited number of datasets.
+    Parameters
+    ----------
+    X: 2-D numpy array of size [n_T, n_V]
+        The data to estimate the dimensionality for PCA.
+        Each column is one feature. Each row is one sample.
+        X is z-scored before further calculation.
+    Returns:
+    --------
+    ncomp: integer
+        The optimal number of components determined by BIC approximation
+        of the likelihood of each candidate number of component
+    log_p: 1-D numpy array
+        log(p(X|k)) for each candidate number of component k, calculated
+        by the BIC approximation provided by Minka 2000.
+    """
+    [n_T, n_V] = X.shape
+    X = scipy.stats.zscore(X)
+    sing = np.linalg.svd(X, full_matrices=False, compute_uv=False)
+    # singular values of data X
+    sing = sing[np.logical_not(np.isclose(sing, 0))]
+    # Remove the singular values that are close to 0.
+    Lambda = sing ** 2
+    # eigen values of the sample covariance matrix.
+    d_max = Lambda.size - 1
+    # Our goal is dimensionality reduction, so at least reducing by 1.
+    K = np.arange(1, d_max + 1)
+    log_p = np.zeros(d_max)
+    # alpha = 2 # A small number corresponding to equation (48) of Minka 2000
+    # n = n_T + 1 + alpha
+    m = Lambda.size * K - K * (K + 1) / 2
+    v_hat = (np.sum(Lambda) - np.cumsum(Lambda[:d_max])) / (Lambda.size - K)
+    log_Lambda = np.log(Lambda)
+    log_p = - n_T / 2 * np.cumsum(log_Lambda[:d_max]) \
+        - n_T * (Lambda.size - K) / 2 * np.log(v_hat) \
+        - (m + K) / 2 * np.log(n_T * n_V)
+    # Equation (78) of Minka 2000. Notice that we operate on the
+    # space spanned by the eigen vectors. Although Minka counted the
+    # number of data point with n_T in this case, we count observations
+    # of all time points and all voxels (n_T * n_V)
```

Comment:
No mathematical justification. See the comment above. Yeah I know this is super hacky.